scripts/scraper-ng/index.js

```diff
@@ -1,51 +1,79 @@
+const fetch = require("node-fetch");
 const rehype = require("rehype");
 const reporter = require("vfile-reporter");
 
+const mdnUrl = require("./mdn-url");
 const toVFile = require("./url-to-vfile");
 
 const examplePage =
   "https://developer.mozilla.org/en-US/docs/Web/HTML/Element/div";
+const exampleShorthand = "/en-US/docs/Web/HTML/Element/div";
 
 const argv = require("yargs")
-  .usage("Usage: $0 <url> …")
-  .example(`$0 ${examplePage}`, "scrape a page")
-  .example(`$0 -n ${examplePage}`, "dry run (lint)")
+  .parserConfiguration({ "boolean-negation": false })
+  .usage("Usage: $0 <url>")
+  .example(`$0 ${examplePage}`, "scrape a page and all its subpages")
+  .example(`$0 ${exampleShorthand}`, "omit the protocol and domain")
+  .example(`$0 --no-subpages ${exampleShorthand}`, "one page only")
+  .example(`$0 -n ${exampleShorthand}`, "dry run (lint)")
   .demandCommand(1, 1, "A URL is required", "Too many arguments")
+  .describe("no-subpages", "Don't walk the entire tree from the URL")
+  .default("no-subpages", false)
   .describe("n", "Dry run (lint)")
   .alias("n", "dry-run")
   .describe("q", "Suppress success messages")
   .alias("q", "quiet")
   .describe("v", "Expand linter notes, if applicable")
   .alias("v", "verbose")
-  .boolean(["n", "q", "v"]).argv;
+  .boolean(["n", "q", "v", "no-subpages"]).argv;
 
 const processor = rehype()
   .use([require("./plugins/kumascript-macros")])
   .use([require("./preset")]);
 // TODO: add YAML frontmatter insertion
 
 async function run() {
-  const vfiles = [];
+  const root = await fetchTree(argv._[0]);
+  const urls = argv.noSubpages ? [root.url] : flattenTree(root);
+  const toProcess = urls.map(toVFile);
+  const processed = [];
 
-  const file = await toVFile(argv._[0]);
+  for (const f of toProcess) {
+    const file = await f;
+    const hasFatalError = file.messages.reduce(
+      (prev, curr) => prev || curr.fatal,
+      false
+    );
```

Comment:
The reason I used `reduce` here is because I know that `vfile` itself can set messages when creating a new instance. So it's possible that there could be messages that weren't created by my `toVFile()` function. To safely set `f.data.hasFatalError`, I'd probably have to do a check like the `reduce` function anyway.

Also, I was thinking ahead a little bit. What if there's metadata we want to check before linting the content? For example, suppose we wanted to lint the presence of certain tags, or use them to direct the linter to follow a particular recipe. It would make sense to set messages for metadata at or immediately after creating the vfile, but before we linted the content. Some of those errors might be fatal (we couldn't fetch the metadata at all) but some we might be able to live with (maybe we couldn't choose a corresponding recipe, but we could still lint for things which are generic to all page types).

But we don't actually need any of that yet, so I'm fine with cutting it until we do. I'll go a step simpler than you've suggested and just check for the presence of any messages on the vfile whatsoever. If we have any messages this early in the process, it probably suggests that linting the page isn't going to be particularly interesting anyway.

Reply:
>  The reason I used reduce here is because I know that vfile itself can set messages when creating a new instance. 

Fair enough. I did wonder about this, but if the messages were fatal, wouldn't it throw? Given that `fail()` throws, that seems likely, and then this wouldn't arise (because you're not catching the constructor call (https://github.com/mdn/stumptown-content/blob/master/scripts/scraper-ng/url-to-vfile.js#L17).

But anyway. This all seems quite speculative, and what you have here seems fine (and what you had before was fine too, really).